src/events/messageCreate/train.ts

```diff
@@ -0,0 +1,21 @@
+import { ArgsOf, Discord, On } from "discordx";
+
+let activeTrain = false;
```

Comment:
Possible issue explained here: https://discord.com/channels/723147134605590598/1175206407998558250/1175950380450984040
but I think activeTrain needs to be per channel id or else other channels will interfere with ongoing trains. Maybe a `Set<channel id>`? where channels currently stored in this set are channels with an active train that was already responded to by the bot (since that state is used to prevent the bot from responding twice to one train).

Reply:
This solution will have an edge case where the bot sends a message from some other command/event ([example](https://github.com/Kyle-Stadelmann/DiscordBot/blob/5706ac7c9095a8eba8c98fad2353ad460e649654/src/events/messageCreate/daniel-ty.ts#L17C4-L17C4)) and then some users start a train on that response, the bot will join in on that train despite already being the first message in that train. So maybe an additional check that one of the first 2 msgs weren't from this `client.botId` is necessary

Reply:
Also another edge case I think will be when 2 users send msgs without string contents (I think this would be when users send solely images or embeds, but im not sure if the images just get converted into a discord cdn link and then embeded like gifs are). So we could either exclude trains where `!content` or also support trains for attachments maybe. Could also make supporting attachments another story for the future

Reply:
All of the wonderful cases dealing with attachments in the Train event:

1. Last 2 users copy paste discord generated link from message.attachment: Already covered
2. User posts image, 2nd user right click and copies _link_ (which is exactly the string message1.attachments[n].attachment) and pastes it which becomes message.content: need to check if message2.content === message1.attachments[n].attachment (or reverse although i assume extremely rare)
    - I expect this to be the most common case, but won't cover it because I'm lazy. If just one more user posts the link, it becomes the case before this one (which means the bot would jump on the train after the 3rd repitition and potentially miss it, cry your heart out)
    - Populates the message.embeds property which is critical information to know for other cases <sup>[1]</sup>
3. User posts image, 2nd user copy pastes the _image_, (of which its message.embeds[n].url is exactly the string message1.attachments[n].attachment) and pastes it: need to check if message2.embed.url === message1.attachments[n].attachment
4. Last 2 users post the same exact file from their clients respectively
    - Discord generates 3 *strictly unique* url's per attachment, *regardless* of the data they contain (in my testing) meaning we cannot exactly compare image likeness
    - message.content of both is empty "": I'm going to cover this
    - A horrible method to implement this, as the images are not garunteed to be the same data-wise, is to (first sort by name since these collections aren't strictly ordered) then compare each message.attachments[n].attachment.name (which is the file name), and if they are the same, append any of the 3 generated urls to the discord bot's message.content which will then embed itself
    - Suprisingly it might actually save work to use a library that calculates image-likeness, so we can hop on a train when they are determined to match visually
5. One user posts a discord generated link (circa case 1,2) and the other copy pastes the _image_ (circa case 3).
    - If you think about it a little you find that you can work this one out with the string data
    - But if you think about it a little harder you realize they all have to be based from the image upload from the original user, so if someone downloads and uploads (the same exact image mind you) then its over <sup>[2]</sup>
    - Another case for using an image-comparison library I guess

- For most the above cases, we must also handle comparing message content itself, meaning each case is likely to be its own helper function called in order if the criteria are met, to keep everything organized
- There's ~~probably~~ definitely more weird cases I'm missing (I came up with like 3 more then forgot them but they're related to these: <sup>[1]</sup><sup>[2]</sup>)
- Aside from the empty string check in the 4th case I'm not doing this stuff in this issue lmao. Seems like a fun project that will take days of banging your head on the wall and would be used once on production if you're lucky. I'll open up an issue linking this comment but i'm assuming we wanna label it "not gonna do but known" or w/e the tag is

Reply:
https://github.com/Kyle-Stadelmann/DiscordBot/issues/47